File: VB/SchedulerHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Windows
Imports System.Windows.Input
Imports DevExpress.Xpf.Scheduler
Imports DevExpress.XtraScheduler

Namespace SchedulerMVVMScenarioWpf
	Public Class SchedulerHelper
		Public Shared ReadOnly AppointmentsModifiedCommandProperty As DependencyProperty = DependencyProperty.RegisterAttached("AppointmentsModifiedCommand", GetType(ICommand), GetType(SchedulerHelper), New PropertyMetadata(Nothing, New PropertyChangedCallback(AddressOf AppointmentsModifiedCommandPropertyChanged)))

		Public Shared ReadOnly FetchAppointmentsCommandProperty As DependencyProperty = DependencyProperty.RegisterAttached("FetchAppointmentsCommand", GetType(ICommand), GetType(SchedulerHelper), New PropertyMetadata(Nothing, New PropertyChangedCallback(AddressOf FetchAppointmentsCommandPropertyChanged)))

		Private Shared commandAppointmentsModified As ICommand = Nothing
		Private Shared commandFetchAppointments As ICommand = Nothing

		Private Shared Sub AppointmentsModifiedCommandPropertyChanged(ByVal obj As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
			Dim schedulerControl As SchedulerControl = TryCast(obj, SchedulerControl)

			If schedulerControl IsNot Nothing Then
				AddHandler schedulerControl.StorageChanged, AddressOf schedulerControl_StorageChanged
				commandAppointmentsModified = CType(e.NewValue, ICommand)
			End If
		End Sub

		Private Shared Sub FetchAppointmentsCommandPropertyChanged(ByVal obj As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
			Dim schedulerControl As SchedulerControl = TryCast(obj, SchedulerControl)

			If schedulerControl IsNot Nothing Then
				AddHandler schedulerControl.StorageChanged, AddressOf schedulerControl_StorageChanged
				commandFetchAppointments = CType(e.NewValue, ICommand)
			End If
		End Sub

		Private Shared Sub schedulerControl_StorageChanged(ByVal sender As Object, ByVal e As EventArgs)
			Dim schedulerControl As SchedulerControl = TryCast(sender, SchedulerControl)
			Dim schedulerStorage As SchedulerStorage = TryCast(schedulerControl.Storage, SchedulerStorage)

			If schedulerStorage IsNot Nothing Then
				AddHandler schedulerStorage.AppointmentsInserted, AddressOf AppointmentsModifiedAction
				AddHandler schedulerStorage.AppointmentsChanged, AddressOf AppointmentsModifiedAction
				AddHandler schedulerStorage.AppointmentsDeleted, AddressOf AppointmentsModifiedAction
				AddHandler schedulerStorage.FetchAppointments, AddressOf FetchAppointmentsAction
			End If
		End Sub

		Private Shared Sub AppointmentsModifiedAction(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs)
			If commandAppointmentsModified IsNot Nothing Then
				commandAppointmentsModified.Execute(Nothing)
			End If
		End Sub

		Private Shared Sub FetchAppointmentsAction(ByVal sender As Object, ByVal e As FetchAppointmentsEventArgs)
			If commandFetchAppointments IsNot Nothing Then
				commandFetchAppointments.Execute(New Object() { (CType(sender, DevExpress.Xpf.Scheduler.Native.SchedulerDataStorage)).Appointments.DataSource, e.Interval })
			End If
		End Sub

		Public Shared Sub SetAppointmentsModifiedCommand(ByVal element As UIElement, ByVal value As ICommand)
			element.SetValue(AppointmentsModifiedCommandProperty, value)
		End Sub

		Public Shared Function GetAppointmentsModifiedCommand(ByVal element As UIElement) As ICommand
			Return CType(element.GetValue(AppointmentsModifiedCommandProperty), ICommand)
		End Function

		Public Shared Sub SetFetchAppointmentsCommand(ByVal element As UIElement, ByVal value As ICommand)
			element.SetValue(FetchAppointmentsCommandProperty, value)
		End Sub

		Public Shared Function GetFetchAppointmentsCommand(ByVal element As UIElement) As ICommand
			Return CType(element.GetValue(FetchAppointmentsCommandProperty), ICommand)
		End Function
	End Class
End Namespace

File: CS/SchedulerHelper.cs
using System;
using System.Windows;
using System.Windows.Input;
using DevExpress.Xpf.Scheduler;
using DevExpress.XtraScheduler;

namespace SchedulerMVVMScenarioWpf {
    public class SchedulerHelper {
        public static readonly DependencyProperty AppointmentsModifiedCommandProperty =
            DependencyProperty.RegisterAttached("AppointmentsModifiedCommand",
            typeof(ICommand), typeof(SchedulerHelper),
            new PropertyMetadata(null, new PropertyChangedCallback(AppointmentsModifiedCommandPropertyChanged)));

        public static readonly DependencyProperty FetchAppointmentsCommandProperty =
            DependencyProperty.RegisterAttached("FetchAppointmentsCommand",
            typeof(ICommand), typeof(SchedulerHelper),
            new PropertyMetadata(null, new PropertyChangedCallback(FetchAppointmentsCommandPropertyChanged)));

        private static ICommand commandAppointmentsModified = null;
        private static ICommand commandFetchAppointments = null;

        private static void AppointmentsModifiedCommandPropertyChanged(DependencyObject obj, DependencyPropertyChangedEventArgs e) {
            SchedulerControl schedulerControl = obj as SchedulerControl;

            if (schedulerControl != null) {
                schedulerControl.StorageChanged += schedulerControl_StorageChanged;
                commandAppointmentsModified = (ICommand)e.NewValue;
            }
        }

        private static void FetchAppointmentsCommandPropertyChanged(DependencyObject obj, DependencyPropertyChangedEventArgs e) {
            SchedulerControl schedulerControl = obj as SchedulerControl;

            if (schedulerControl != null) {
                schedulerControl.StorageChanged += schedulerControl_StorageChanged;
                commandFetchAppointments = (ICommand)e.NewValue;
            }
        }

        private static void schedulerControl_StorageChanged(object sender, EventArgs e) {
            SchedulerControl schedulerControl = sender as SchedulerControl;
            SchedulerStorage schedulerStorage = schedulerControl.Storage as SchedulerStorage;

            if (schedulerStorage != null) {
                schedulerStorage.AppointmentsInserted += AppointmentsModifiedAction;
                schedulerStorage.AppointmentsChanged += AppointmentsModifiedAction;
                schedulerStorage.AppointmentsDeleted += AppointmentsModifiedAction;
                schedulerStorage.FetchAppointments += FetchAppointmentsAction;
            }
        }

        private static void AppointmentsModifiedAction(object sender, PersistentObjectsEventArgs e) {
            if (commandAppointmentsModified != null)
                commandAppointmentsModified.Execute(null);
        }

        private static void FetchAppointmentsAction(object sender, FetchAppointmentsEventArgs e) {
            if (commandFetchAppointments != null)
                commandFetchAppointments.Execute(new object[] { 
                    ((DevExpress.Xpf.Scheduler.Native.SchedulerDataStorage)sender).Appointments.DataSource, e.Interval 
                });
        }

        public static void SetAppointmentsModifiedCommand(UIElement element, ICommand value) {
            element.SetValue(AppointmentsModifiedCommandProperty, value);
        }

        public static ICommand GetAppointmentsModifiedCommand(UIElement element) {
            return (ICommand)element.GetValue(AppointmentsModifiedCommandProperty);
        }

        public static void SetFetchAppointmentsCommand(UIElement element, ICommand value) {
            element.SetValue(FetchAppointmentsCommandProperty, value);
        }

        public static ICommand GetFetchAppointmentsCommand(UIElement element) {
            return (ICommand)element.GetValue(FetchAppointmentsCommandProperty);
        }
    }
}